daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/CheckAuthorization.scala

```diff
@@ -0,0 +1,118 @@
+
+package com.daml.lf
+package speedy
+
+import com.daml.lf.ledger.Authorize
+import com.daml.lf.ledger.FailedAuthorization
+import com.daml.lf.transaction.Node.{NodeCreate, NodeFetch, NodeLookupByKey}
+
+import PartialTransaction.ExercisesContext
+
+private[lf] object CheckAuthorization {
+
+  def authorize(
+      passIf: Boolean,
+      failWith: => FailedAuthorization,
+  ): List[FailedAuthorization] = {
+    if (passIf)
+      List()
+    else
+      List(failWith)
+  }
+
+  def authorizeCreate(
+      create: NodeCreate[_, _],
+      auth: Authorize,
+  ): List[FailedAuthorization] = {
+    authorize(
+      passIf = create.signatories subsetOf auth.authParties,
+      failWith = FailedAuthorization.CreateMissingAuthorization(
+        templateId = create.coinst.template,
+        optLocation = create.optLocation,
+        authorizingParties = auth.authParties,
+        requiredParties = create.signatories,
+      ),
+    ) ++
+      authorize(
+        passIf = create.signatories.nonEmpty,
+        failWith = FailedAuthorization.NoSignatories(create.coinst.template, create.optLocation),
+      ) ++
+      (create.key match {
+        case None => List()
+        case Some(key) =>
+          val maintainers = key.maintainers
+          authorize(
+            passIf = maintainers subsetOf create.signatories,
+            failWith = FailedAuthorization.MaintainersNotSubsetOfSignatories(
+              templateId = create.coinst.template,
+              optLocation = create.optLocation,
+              signatories = create.signatories,
+              maintainers = maintainers,
+            ),
+          )
+      })
+  }
+
+  def authorizeFetch(
+      fetch: NodeFetch[_, _],
+      auth: Authorize,
+  ): List[FailedAuthorization] = {
+    authorize(
+      passIf = fetch.stakeholders.intersect(auth.authParties).nonEmpty,
+      failWith = FailedAuthorization.FetchMissingAuthorization(
+        templateId = fetch.templateId,
+        optLocation = fetch.optLocation,
+        stakeholders = fetch.stakeholders,
+        authorizingParties = auth.authParties,
+      )
+    )
+  }
+
+  def authorizeLookupByKey(
+      lbk: NodeLookupByKey[_, _],
+      auth: Authorize,
+  ): List[FailedAuthorization] = {
+    authorize(
+      passIf = lbk.key.maintainers subsetOf auth.authParties,
+      failWith = FailedAuthorization.LookupByKeyMissingAuthorization(
+        lbk.templateId,
+        lbk.optLocation,
+        lbk.key.maintainers,
+        auth.authParties,
+      )
+    )
+  }
+
+  def authorizeExercise(
+      ex: ExercisesContext,
+      auth: Authorize,
+  ): List[FailedAuthorization] = {
+    val controllersDifferFromActors = ex.controllers != ex.actingParties
+    authorize(
+      passIf = ex.actingParties.nonEmpty,
+      failWith = FailedAuthorization.NoControllers(ex.templateId, ex.choiceId, ex.optLocation),
+    ) ++
+      authorize(
+        passIf = !controllersDifferFromActors, // TODO: can this failure ever occur?
```

Comment:
The construct that can trigger this does still exist in DAML-LF although the compiler stopped producing it a while ago.
```suggestion
        passIf = ex.controllers == ex.actingParties,
```

Reply:
ok, thanks. I remove the `TODO` comment